File: Leetcode 0098 Validate Binary Search Tree [medium] [tree]/Validate Binary Search Tree [divide-conquer].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    return isValid(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};

let isValid = (root, min, max) => {
    if (root === null) {
        return true;
    }
    
    if (root.val >= max || root.val <= min) {
        return false;
    }
    
    return isValid(root.left, min, Math.min(max, root.val)) && isValid(root.right, Math.max(root.val, min), max);
}

File: Leetcode 0098 Validate Binary Search Tree [medium] [tree]/Validate Binary Search Tree [divide-conquer].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    
    private boolean isValidBST(TreeNode root, long min, long max) {
        if (root == null) {
            return true;
        }
        
        if (root.val <= min || root.val >= max) {
            return false;
        }
        
        // 左子樹上界合理範圍: [min + 1, root.val - 1]
        // 右子樹下界合理範圍: [root.val + 1, max - 1]
        return isValidBST(root.left, min, Math.min(root.val, max)) && isValidBST(root.right, Math.max(root.val, min), max);
    }
}